### PR TITLE
rmp: Adds support to create an abc standard cell library from opensta

### PR DESCRIPTION
**Pull request by QuantamHD**
Creates a factory class to map an sta library into an abc one. This library can be used to unmap a netlist in ABC into an AIG for optimization, and then back into a mapped netlist

### CODE REVIEW

File: src/rmp/src/abc_library_factory.cpp [hunk 1]
```diff
@@ -0,0 +1,468 @@
...
+      continue;
+    }
+
+    if (HasNonInputOutputPorts(cell)) {
+      continue;
+    }
+
+    if (HasOnlyPowerAndGroundPorts(cell)) {
+      continue;
+    }
```
Comment (maliberty) [inline src/rmp/src/abc_library_factory.cpp:379]: Would it be simpler to just have IsAbcCompatible and traverse the pins once?

Comment (QuantamHD) [reply src/rmp/src/abc_library_factory.cpp:379]: I feel like that will end up being less clear overall since you need to traverse all the pins for some of the checks so you'll end up with a while loop with a bunch of counters that you then need to check. I feel like these checks could be moved into IsAbcCompatible, but I'd rather keep them seperate.

File: src/rmp/src/abc_library_factory.cpp [hunk 2]
```diff
@@ -0,0 +1,468 @@
...
+#include "sta/PortDirection.hh"
+#include "sta/Sta.hh"
+#include "sta/TableModel.hh"
+#include "sta/TimingArc.hh"
+#include "sta/Units.hh"
+#include "utl/deleter.h"
+
+namespace rmp {
+
+static bool IsCombinational(sta::LibertyCell* cell)
```
[the remark this answers is not in the document]
Comment (QuantamHD) [reply src/rmp/src/abc_library_factory.cpp:31]: No that function deals with insts not masters which is what we have here. The function looks at net source type which is not meaningful in this context. If you create a version that worked on masters I could use. I transitioned everything over to dbSta so it wouldn't be hard to change.

[the remark this answers is not in the document]
Comment (maliberty) [reply src/rmp/src/abc_library_factory.cpp:31]: It makes finer distinction that I guess you care about that do require an instance.  If you stick with this scheme you should consider isLevelShifter and isIsolationCell.

File: src/rmp/src/abc_library_factory.cpp (hunk 1 above)
Comment (maliberty) [reply src/rmp/src/abc_library_factory.cpp:379]: If you have an output, how can you have only power and ground ports?

Comment (QuantamHD) [reply src/rmp/src/abc_library_factory.cpp:379]: They are considered inout ports

File: src/rmp/src/abc_library_factory.cpp [hunk 3]
```diff
@@ -0,0 +1,468 @@
...
+// https://developers.google.com/open-source/licenses/bsd
+
+#include "abc_library_factory.h"
+
+#include <cmath>
+#include <unordered_set>
+#include <vector>
+
+// Poor include definitions in ABC
+// clang-format off
```
[the remark this answers is not in the document]
Comment (QuantamHD) [reply src/rmp/src/abc_library_factory.cpp:16]: Yes.

**Comment by QuantamHD** [comment]
Needs https://github.com/berkeley-abc/abc/pull/294 to be merged here and then in the openroad copy

File: src/rmp/src/abc_library_factory.cpp (hunk 1 above)
Comment (maliberty) [reply src/rmp/src/abc_library_factory.cpp]: inout are not outputs unless you mean to use isAnyOutput() not isOutput()

Comment (maliberty) [reply src/rmp/src/abc_library_factory.cpp]: Also they are not considered inout either but have their own categories.

**Comment by QuantamHD** [comment]
@maliberty Can you merge this

File: src/rmp/src/CMakeLists.txt [hunk 4]
```diff
@@ -73,6 +73,23 @@ target_link_libraries(rmp
...
  )
 
+add_library(rmp_abc_library 
+  abc_library_factory.cpp
+)
+
+target_include_directories(rmp_abc_library
+  PUBLIC
+    ../include
+    .
```
[the remark this answers is not in the document]
Comment (QuantamHD) [reply src/rmp/src/CMakeLists.txt:84]: Done.

File: src/rmp/src/abc_library_factory.cpp (hunk 1 above)
Comment (QuantamHD) [reply src/rmp/src/abc_library_factory.cpp]: You're right. I removed it.

File: src/dbSta/src/MakeDbSta.cc [hunk 5]
```diff
@@ -29,11 +29,6 @@ namespace ord {
 
 using sta::dbSta;
 
-dbSta* makeDbSta()
```
Comment (maliberty) [inline src/dbSta/src/MakeDbSta.cc:32]: Why is this moved?

Comment (QuantamHD) [reply src/dbSta/src/MakeDbSta.cc:32]: In order to construct this from tests the linker needs to see this function in the lib version of dbsta instead of the binary version. 

The new operator itself isn't callable directly because of the abstract gui class pointers.

File: CMakeLists.txt [hunk 6]
```diff
@@ -140,6 +140,7 @@ configure_file(
 
 ################################################################
 
+link_libraries(stdc++fs)
```
Comment (maliberty) [inline CMakeLists.txt:143]: I tried this previously but had lots of trouble on mac.  Hopefully this will work ok a few years later.

**Review by maliberty — APPROVED** [review]

File: CMakeLists.txt (hunk 6 above)
Comment (QuantamHD) [reply CMakeLists.txt:143]: C++ makes me sad
